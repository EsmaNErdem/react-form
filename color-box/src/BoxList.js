import React, { useState } from "react";
import {v4 as uuid} from "uuid";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";

const BoxList = () => {
    const INITIAL_STATE = [
        {id: uuid(), color:"greenyellow", width:"200px", height: "90px"},
        {id: uuid(), color:"deeppink", width:"150px", height: "105px"}
    ]

    const [boxes, setBoxes] = useState(INITIAL_STATE);

    const addBox = (newBox) => {
        setBoxes(boxes => [...boxes, {...newBox, id:uuid()}])
    }

    const removeBox = (id) => {
        setBoxes(boxes => boxes.filter(b => b.id !== id))
    }


    return (
        <div className="BoxList">
            <NewBoxForm addBox={addBox}/>
            <div>
                {boxes.map(({id, color, width, height}) => (
                <Box 
                id={id}
                key={id} 
                color={color} 
                width={width} 
                height={height}
                handleRemove={removeBox}
                />))}
            </div>
        </div>
    )
}

export default BoxList;