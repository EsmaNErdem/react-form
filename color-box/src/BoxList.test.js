import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

test("renders without crashing", () => {
    render(<BoxList />);
});

test("it renders and matches with snaphot", () => {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});


it("should add a new box", () => {
    const { queryByText, getByLabelText, queryByTestId, getAllByDisplayValue } = render(<BoxList />);
    const colorInput = getByLabelText("Background Color:");
    const widthInput = getByLabelText("Width:");
    const heightInput = getByLabelText("Height:");
    const btn = queryByText("Add a new box!");
// ensure that remove button for the box is not there yet
    expect(queryByTestId('orangered')).not.toBeInTheDocument();
    fireEvent.change(colorInput, { target: { value: 'orangered' } });
    fireEvent.change(widthInput, { target: { value: '200px' } });
    fireEvent.change(heightInput, { target: { value: '100px' } });
    fireEvent.click(btn);
// make sure the remove button is there now
    const removeButton = queryByTestId('orangered');
    expect(removeButton).toBeInTheDocument();

    expect(removeButton.previousSibling).toHaveStyle(`
    width: 200px;
    height: 100px;
    background: orangered;
     `);

     // expect form to be empty
    expect(getAllByDisplayValue("")).toHaveLength(3);
})

it("should remove a box", () => {
    const { queryByTestId } = render(<BoxList />);

    const removeButton = queryByTestId('deeppink');
    expect(removeButton).toBeInTheDocument();
    expect(removeButton.previousSibling).toHaveStyle(`
    background: deeppink; 
    width: 150px; 
    height: 105px;
    `);

    fireEvent.click(removeButton);
    expect(removeButton).not.toBeInTheDocument();
    expect(removeButton.previousSibling).toBeNull()
})