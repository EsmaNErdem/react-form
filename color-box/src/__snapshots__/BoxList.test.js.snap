// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snaphot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <form>
      <label
        for="color"
      >
        Background Color:
      </label>
      <input
        id="color"
        name="color"
        type="text"
        value=""
      />
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        type="text"
        value=""
      />
      <button>
        Add a new box!
      </button>
    </form>
    <div>
      <div
        class="Box"
        style="background: greenyellow; width: 200px; height: 90px;"
      />
      <button
        data-testid="greenyellow"
      >
        X
      </button>
      <div
        class="Box"
        style="background: deeppink; width: 150px; height: 105px;"
      />
      <button
        data-testid="deeppink"
      >
        X
      </button>
    </div>
  </div>
</DocumentFragment>
`;
